src/librustc/traits/error_reporting.rs

```diff
@@ -357,14 +361,20 @@ pub fn report_selection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
                             let trait_ref = trait_predicate.to_poly_trait_ref();
                             let mut err = struct_span_err!(
                                 infcx.tcx.sess, obligation.cause.span, E0277,
-                                "the trait `{}` is not implemented for the type `{}`",
-                                trait_ref, trait_ref.self_ty());
-
-                            // Check if it has a custom "#[rustc_on_unimplemented]"
-                            // error message, report with that message if it does
-                            let custom_note = report_on_unimplemented(infcx, &trait_ref.0,
-                                                                      obligation.cause.span);
-                            if let Some(s) = custom_note {
+                                "the predicate `{}` is not satisfied",
+                                trait_ref.to_predicate());
+
+                            // Try to report a good error message.
```

Comment:
It'd be nice to be a bit more specific than this comment. For example, put in the body of each if:

`if ... {
    // Prefer to suggest an actionable fix `"add `T: Foo`"
} else {
    // Otherwise, custom explanation if available
}
```

At first, I was not sure if this should be an "xor" sort of logic -- but now I agree with you. So long as there is a chance that this is just a missing bound, it seems wrong to give a message like "you can't index a vector with an `X`" or whatever, since it may well be that you *could*. Might be worth adding (in the comment) your reasoning as to why it makes sense to give preference the actionable suggestion. (I am guessing as to your motivation here, maybe I have it wrong).

Basically these kind of "soft logic" decisions are really easy to lose track of, so leaving a detailed comment trail let's us refine them in the future without fear of overlooking use cases we have forgotten about.